### PR TITLE
(closes #131) Embed YAML-LD into HTML `<script>` tag

### LINKED ISSUE #131: YAML under `<script>` HTML tag unsupported in the spec
JSON-LD API specification supports parsing of JSON-LD documents included in HTML documents with `<script>` tag. There is a [test manifest](https://github.com/w3c/json-ld-api/blob/main/tests/html-manifest.jsonld) which tests precisely that.

The YAML-LD spec does not mention anything about that though it does specify the MIME type which could be used for the `<script>` tag.

Do you think we should be supporting that feature?

Note that it will likely conflict with #119: we cannot do both.

### PR DESCRIPTION
**Pull request by anatoly-scherbakov**
We should be able to embed YAML-LD scripts into HTML documents, as described in #131.

***
<a href="https://pr-preview.s3.amazonaws.com/json-ld/yaml-ld/pull/132.html" title="Last updated on Feb 7, 2024, 5:08 PM UTC (10a8502)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/json-ld/yaml-ld/132/efd92b8...10a8502.html" title="Last updated on Feb 7, 2024, 5:08 PM UTC (10a8502)">Diff</a>

### CODE REVIEW

File: spec/index.html [hunk 1]
```diff
@@ -1007,6 +1007,39 @@ <h2>Interoperability Considerations</h2>
...
+    "@context": https://json-ld.org/contexts/person.jsonld
+    "@id": http://dbpedia.org/resource/Cynthia_Lennon
+    born: 1939-09-10
+    </script>
+    -->
+    </pre>
+
+    <p data-tests="manifest.html#html-and-yaml-streams">
+      If the YAML-LD <code>&lt;script></code> tag contains a <a>YAML Stream</a> with multiple <a>YAML documents</a>, each of these documents MUST be treated as if it was included in a separate <code>&lt;script></code> tag.
+    </p>
```
Comment (gkellogg) [inline spec/index.html:1040]: I've done some updates to parse a YAML stream from an HAML script, but I need to un-indent the script block, as the leading whitespace on the document separator causes the YAML parser I use to fail. We should discuss the need to unindent the script contents.

**Review by gkellogg — CHANGES_REQUESTED** [review]
Note the comment about the need to unindent a stream. Probably not necessary for documents in general.

File: spec/index.html (hunk 1 above)
Comment (anatoly-scherbakov) [reply spec/index.html:1055]: Added notes & a test about that.

File: spec/index.html [hunk 2]
```diff
@@ -1007,6 +1007,54 @@ <h2>Interoperability Considerations</h2>
...
+    </script>
+    -->
+    </pre>
+
+    <p data-tests="manifest.html#html-dedent-needed,manifest.html#html-dedent-not-needed">
+      YAML syntax is indentation based. Before processing each `&lt;script>` block with YAML-LD content, YAML-LD processor MUST <em>dedent</em> the body of the block using the following procedure:
+    </p>
+
+    <ul>
+      <li>Count the number of <strong>leading spaces</strong> in the first line of the YAML-LD block, and save that to `dedent_count` variable,</li>
```
Comment (gkellogg) [inline spec/index.html:1043]: Maybe "first non-empty line", as some markup may have an empty line before actual YAML content starts.

**Review by gkellogg — APPROVED** [review]

File: spec/index.html (hunk 2 above)
Comment (anatoly-scherbakov) [reply spec/index.html]: Added!

**Comment by anatoly-scherbakov** [comment]
@gkellogg I do recall that we probably can't merge PRs to the spec right now. Should we keep these PRs open?

Or maybe we should discuss this on the upcoming CG meeting?

**Comment by gkellogg** [comment]
Yes, I think we need to keep this open. Let's discuss how we move forward on the next call, but I think it requires a WG action and probably needs to await the outcome of an updated WG charter.

File: spec/index.html [hunk 3]
```diff
@@ -1007,6 +1007,54 @@ <h2>Interoperability Considerations</h2>
...
+    </script>
+    -->
+    </pre>
+
+    <p data-tests="manifest.html#html-dedent-needed,manifest.html#html-dedent-not-needed">
+      YAML syntax is indentation based. Before processing each `&lt;script>` block with YAML-LD content, YAML-LD processor MUST <em>dedent</em> the body of the block using the following procedure:
+    </p>
+
+    <ul>
+      <li>Count the number of <strong>leading spaces</strong> in the first non-empty line of the YAML-LD block, and save that to `dedent_count` variable,</li>
```
Comment (TallTed) [inline spec/index.html:1043]:
```suggestion
      <li>Count the number of <strong>leading spaces</strong> in the first non-empty line of the YAML-LD block, and save that to `dedent_count` variable.</li>
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Merging this should trigger a run of PR-Preview, which should now succeed

**Comment by gkellogg** [comment]
The JSON-LD WG discussed this today and suggested that work on the YAML-LD editor's draft can proceed.

**Comment by VladimirAlexiev** [comment]
@anatoly-scherbakov 
> YAML syntax is indentation based. 

Can you make this more forceful to say something like "leave spaces and newlines alone" or "don't mangle spaces and newlines".

> Before processing each  <script>  block with YAML-LD content, YAML-LD processor  MUST  dedent  the body of the block using the following procedure: 

Afaik, no such dedenting is needed. I loaded your example to [YAML Playground](https://play.yaml.io/main/parser?input=ICAiQGNvbnRleHQiOiBodHRwczovL2pzb24tbGQub3JnL2NvbnRleHRzL3BlcnNvbi5qc29ubGQKICAiQGlkIjogaHR0cDovL2RicGVkaWEub3JnL3Jlc291cmNlL0pvaG5fTGVubm9uCiAgbmFtZTogSm9obiBMZW5ub24KICBib3JuOiAxOTQwLTEwLTA5CiAgc3BvdXNlOiBodHRwOi8vZGJwZWRpYS5vcmcvcmVzb3VyY2UvQ3ludGhpYV9MZW5ub24KLS0tCiAgIkBjb250ZXh0IjogaHR0cHM6Ly9qc29uLWxkLm9yZy9jb250ZXh0cy9wZXJzb24uanNvbmxkCiAgIkBpZCI6IGh0dHA6Ly9kYnBlZGlhLm9yZy9yZXNvdXJjZS9DeW50aGlhX0xlbm5vbgogIGJvcm46IDE5MzktMDktMTA=) and played with the global indentation. Look at the parse events in the "New Reference Parser (JS)" pane: they remain the same.

**Comment by VladimirAlexiev** [comment]
@gkellogg (cc @TallTed)
> Note the comment about the need to unindent a stream. Probably not necessary for documents in general.

Yes! See my comment above. More explicitly:
- the stream separator `---` must be at beginning of line
- the global indentation of the docs does not matter.

This version of @anatoly-scherbakov's example returns the same parse events as the original version:
```yaml
  "@context": https://json-ld.org/contexts/person.jsonld
  "@id": http://dbpedia.org/resource/John_Lennon
  name: John Lennon
  born: 1940-10-09
  spouse: http://dbpedia.org/resource/Cynthia_Lennon
---
  "@context": https://json-ld.org/contexts/person.jsonld
  "@id": http://dbpedia.org/resource/Cynthia_Lennon
  born: 1939-09-10
```

So @anatoly-scherbakov I don't think we should require HTML parsers to dedent scripts.
We should only:
- require them to leave newlines and spaces alone
- advise users that it's best for the global indent to be 0, and that is required for the stream separator `---`

**Comment by VladimirAlexiev** [comment]
@anatoly-scherbakov I have a silly concern, but let's honor the life of a great man:
John was married to Cynthia for just 6 years, then married Yoko for the rest of his life.

So can you change the spouse to:
http://dbpedia.org/resource/Yoko_Ono (born 1933-02-18)

**Comment by TallTed** [comment]
> YAML syntax is indentation based.

Syntactically-important, human-invisible whitespace is evil. I cannot tell you how long it has sometimes taken to clean up Oracle configuration files, when one person indented some lines with spaces, and another person indented with tabs.

I am glad that later comments suggest that whitespace is less important ("don't think we should require HTML parsers to dedent scripts"), but "[requiring] them to leave newlines and spaces alone" seem to say that *some* whitespace *is* important?

This is confusing. But maybe it's just this conversation that's confusing, and the docs will be clear, in the end?

**Comment by anatoly-scherbakov** [comment]
Yes, I have revised my own https://github.com/iolanta-tech/python-yaml-ld implementation and it works without dedenting anything. I do not, by this moment, remember why I wanted to dedent :thinking: 

Creating new issues for these. Thank you for pointing them out!

**Pull request merged**